client/app/components/visualizations/editor/TabbedEditor.jsx

```diff
@@ -0,0 +1,37 @@
+import { isFunction, map, filter, merge } from 'lodash';
+import React from 'react';
+import PropTypes from 'prop-types';
+import Tabs from 'antd/lib/tabs';
+import { EditorPropTypes } from '@/visualizations';
+
+export default function TabbedEditor({ tabs, options, data, onOptionsChange, ...restProps }) {
+  const optionsChanged = (newOptions, updateFunction = merge) => {
```

Comment:
You're right - interesting and definitely looks weird. Added some abstraction for it https://github.com/getredash/redash/pull/4382/commits/bd89634ac42cc0ce6dc42b6e7e8e0509d39bffea

Reply:
I tend to like JS simplicity with its weirdness, also assumed the Lodash functions as a "basic set" in the codebase, thus assuming people (most of them at least) will know it or have the docs open. So, for me it was cool anyway.

Reply:
I added this little abstraction not because I think somebody may not know how `_.merge` or `_.extend` works, but because somebody may not know how they're used in this particular context. The code code like `onOptionsChange({...}, extend)` raises too much questions: why `extend`? What else could be there? What it extends at all? etc. With something like `onOptionsChange({ ... }, UpdateOptionsStrategy.shallowMerge)` it's easy to understand that second argument defines how first argument will be used to modify options; also, no need to know how `shallowMerge` is internally implemented - you could think about it as a black box, and also you'll know that namespace `UpdateOptionsStrategy` contains other constants that could be used instead of it.

Reply:
I see, considering Visualizations is one of the biggest extension points, it makes sense to leave it as simple and comprehensible as possible. so it's a good abstraction :+1:.